### ISSUE TITLE
An API for the Feed to custom scrapers

### ISSUE THREAD

**Issue by tommeagher** [themarshallproject/klaxon#46]
Klaxon is about making it dead simple for someone to watch a basic HTML page, the kind many government websites still use. The feed, the alerts and the snapshot comparisons are all very useful for this. But I could imagine some people might like to add more complex scrapers of their own devising but still make use of the alerts and the feeds. 
What would this entail?

**Comment by jacksongs** [comment]
There's an Australian platform called morph.io that automatically runs scrapers (in Docker containers) every day. I have set up change detection style scrapers there in the past.

I'm not saying it can be used by Klaxon, but might be useful during the design phase.

**Comment by zzolo** [comment]
I think this is a good idea.  A simple API that one could throw content at would be super helpful to handle scraping and content that is a more complicated, while still getting the interface and alerts.

**Comment by tommeagher** [comment]
@zzolo I'm really into this idea. I'm not sure what it would look like yet, but I'd love to see someone willing to tackle this.

**Comment by ivarvong** [comment]
Is the idea that instead of downloading a page's HTML from Klaxon, you'd instead download with some other script, managed independently, then push the results to Klaxon to figure out if something has changed?

**Comment by zzolo** [comment]
Yep

**Comment by zzolo** [comment]
@ivarvong I've gotten token based auth for the API going, but I am having trouble at how to call a compare task/method if someone send along an id and content.  If you could just point me what I need to call, that would help.

**Comment by tommeagher** [comment]
I was thinking more that Klaxon could be split into two pieces: 
1. The alerts (via email, slack, etc) and the Feed
2. The scraping and diffing of simple pages based on css selector

I figured the first part could be an API to receive from custom scrapers, in addition to the stock Klaxon scrapers, any update and send it through to the feed and the alerts system. But if I understand what you're doing, @zzolo, that may be smarter, to keep the archives in Klaxon and handle the diffing there rather than in a separate script.

**Comment by zzolo** [comment]
Yeah, so the idea would be that the Page would be setup in Klaxon, and then via the API, content would be sent to it.  Klaxon would then determine if there is a change, and then notify whoever based on the Page settings.

I think I found the method I should run for the main API call.
https://github.com/themarshallproject/klaxon/blob/f19c42f9efb93397ca5800e539509bd16f96ffbf/app/models/poll_page.rb#L2

(I need to find a way to turn off the scraping for a page by Klaxon itself as well.)

**Comment by GabeIsman** [comment]
@zzolo Yep, that's the right logic. I think there is probably some refactoring in order. Notably we need to have the "download the page, extract the selected text" logic separate from the "compare this version of extracted content with previous versions, save snapshot and notify if necessary".

Once that refactoring is done, I think adding a `custom_scraper` flag to the Pages table makes good sense.

**Comment by zzolo** [comment]
Created a pull request.  Again, my expertise in Ruby or Rails is near none, so it might not be the best way to do this.

**Comment by tommeagher** [comment]
Hi all. With the recent release of [Klaxon Cloud](https://www.documentcloud.org/app?q=#add-ons/MuckRock/Klaxon), we're going back and revisiting old issues that we're not going to pursue or support as we consider any future development of the original standalone Klaxon. This one (almost 8 years old now) falls in that bucket. Thanks for the contributions and discussions on this, but we'll close it as WONTFIX.
